WIP

spencerkimball opened the pull request.

spencerkimball committed:
COMPONENT TRACING WIP

Add support for recording "events" besides spans.

Events are simpler - just counters, no sampling, no log messages, no
nothing. They do optionally have a single error per event counter,
though.
The idea is that sometimes the work that a component does for an
operation is very short in nature and there's little point in creating a
span for it only to close it immediately. For example, the pgwire
component accepts new connections. We want to record the fact that a
connection was accepted, but there's nothing to trace for this
"operation".

System component UI data query revamp

Split single call to component activity into four RPCs:
- Components: fetch counters across all nodes for component activity
- SampleTraces: sample all component traces across all nodes for configurable duration
- ComponentTraces: fetch component traces for a node/component combination after tracing
- ComponentTrace: fetch all samples from all nodes for a trace id after tracing

These changes mean that the system component heat map view can cheaply run as an
info radiator and the fetching of component counts across nodes is de-coupled from
distributed tracing, which is expensive.

Once distributed tracing is requested, it runs for a specified duration and target
count of samples per component. All traces are stored locally at each node instead of
being sent back over the network, which could otherwise be prohibitively expensive.
Calls to ComponentTraces and ComponentTrace reference the data stored locally in
order to send back to the UI only the data necessary to render a view.

TODO: store distributed trace information on temporary persistent storage instead
of in memory and handle garbage collection.

admin ui - fix bugs, add test_ui parameter

Fixed various bugs with displaying "real" cluster data.

Made display of real cluster data the default. The test UI data can be loaded
by specifying "&test_ui" in the URL.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35774)
<!-- Reviewable:end -->

spencerkimball committed:
Fixed errors introduced with use of Longs in test UI

spencerkimball committed:
admin ui: Properly set up spans for node liveness

Also fixed assorted bugs in the UI code, and dealt with spans which are still
pending at time of collection.

Release note: None

spencerkimball committed:
If parent span id is 0, treat as a root for trace display

spencerkimball committed:
admin ui: change component activity refresh period, fix Long bugs

Release note: None

spencerkimball committed:
storage: add support for propagation of trace data to Raft

Release note: None

spencerkimball committed:
storage: expand write batch by decoding it
admin ui: add precise nanosecond times to expanded sample html titles
admin ui: pick out exact component span instead of assuming first span is component span
admin ui: fix bug in overlap algorithm for spans so spans whose duraton exceeds parent overlap

Release note: None

spencerkimball committed:
admin ui: add expandable detail area in sample traces for tags

Enclosed in a bordered div with horizontal and vertical scroll for overflow.
Can be toggled expanded or collapsed.

Release note: None

spencerkimball committed:
admin ui: display rates of component activity instead of counts

Release note: None

spencerkimball committed:
storage: add liveness records as SetTag() calls

Release note: None

andreimatei committed:
add component col to crdb_internal.session_trace

spencerkimball committed:
Fix a couple of bugs in ui

spencerkimball committed:
admin ui: fixes to system components UI

Just use span tags instead of introducing additional attributes.
Numerous fixes for span embedding, including disallowing embedding
across node boundaries and adding estimated durations for pending
spans.

Added system component spans around store send and node status
reporting.

Release note: None

spencerkimball committed:
tsdb: add component spans

- Pruning
- Rollups
- Polling

Release note: None

spencerkimball committed:
storage: add component spans for several new system components

- Range lease acquire (with mark stuck)
- Range lease transfer
- Replica raft write evaluation

Release note: None

spencerkimball committed:
storage: add component spans for txn wait queues

- txn wait push
- txn wait query
- txn resolve contention

Release note: None

spencerkimball committed:
storage: add component span support to storage queues

spencerkimball committed:
rpc: added component spans for external server endpoints

- Admin
- Status
- Authentication
- Time series server

Release note: None

spencerkimball committed:
admin ui: system component fixes

- Moved sample traces controls out of cells and into main UI
- Skip duplicate spans
- Properly set depth of embedded children to parent's depth
- Sort system component names lexicographically
- Display span and parent span IDs for debugging

Release note: None

andreimatei committed:
tolerate nil err in FinishWithErr. Besides being sane, this fixes a crash.

andreimatei committed:
fix a bug - we were wrapping a nil error impl inside a non-nil err
interface

Release note: None

andreimatei committed:
add the batch as tag to the DistSender components

andreimatei committed:
set txn as tag for client.dist.sender.send

andreimatei committed:
cleanup

andreimatei committed:
tags and errors from the span refresher

andreimatei committed:
new sql statements component

andreimatei committed:
add EventLogs

andreimatei committed:
fix a deadlock

- locks were taken in reverse order by:
  - StartChildSpan
    - locks the parent and does s.enableRecording(recordingGroup) - which does spanGroup.Add(s)

versus
  StartComponentSpan - parentRecordingHasComponentSpan
    - locks the recordingGroup then locks the span in the recording group

Release note: None

spencerkimball committed:
admin ui: scroll to component being accessed

Also, fix test_ui view.

Release note: None

andreimatei committed:
better event log for connections

andreimatei committed:
event log in conn executor

andreimatei committed:
render all the event counters in a component tooltip

Release note: None

andreimatei committed:
reintroduce component tracing in Raft applications

andreimatei committed:
debug

andreimatei committed:
fix

andreimatei committed:
put back more raft application tags

andreimatei committed:
react fixes after rebase

andreimatei committed:
fix

bobvawter commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=35774) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=35774) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: spencerkimball<br/>:x: andreimatei<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=35774) it.</sub>
